File: app/src/main/java/com/vikho305/isaho220/outstanding/ui/view/LoginActivity.java
package com.vikho305.isaho220.outstanding.ui.view;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;

import com.vikho305.isaho220.outstanding.R;
import com.vikho305.isaho220.outstanding.ui.viewmodel.ContextualViewModelFactory;
import com.vikho305.isaho220.outstanding.ui.viewmodel.LoginViewModel;
import com.vikho305.isaho220.outstanding.util.Resource;

public class LoginActivity extends AppCompatActivity implements View.OnClickListener {
    private static final int REGISTER_REQUEST = 0;

    private EditText usernameInput, passwordInput;
    private Button registerButton, loginButton;
    private ProgressBar progressBar;

    private LoginViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        initViewModel();

        usernameInput = findViewById(R.id.loginUsername);
        passwordInput = findViewById(R.id.loginPassword);
        registerButton = findViewById(R.id.loginRegisterBtn);
        loginButton = findViewById(R.id.loginLoginBtn);
        progressBar = findViewById(R.id.loginProgressBar);

        registerButton.setOnClickListener(this);
        loginButton.setOnClickListener(this);
    }

    private void initViewModel() {
        ContextualViewModelFactory contextualViewModelFactory = new ContextualViewModelFactory(this);
        viewModel = new ViewModelProvider(this, contextualViewModelFactory).get(LoginViewModel.class);
        viewModel.getAuthToken().observe(this, new Observer<Resource<String>>() {
            @Override
            public void onChanged(Resource<String> stringResource) {
                switch (stringResource.getStatus()) {
                    case SUCCESS:
                        // TODO: start map activity and clear back stack
                        break;
                    case LOADING:
                        progressBar.setVisibility(View.VISIBLE);
                        registerButton.setEnabled(false);
                        loginButton.setEnabled(false);
                        break;
                    case ERROR:
                        progressBar.setVisibility(View.INVISIBLE);
                        registerButton.setEnabled(true);
                        loginButton.setEnabled(true);
                        break;
                }
            }
        });
    }

    @Override
    public void onClick(View v) {
        if (v == registerButton) {
            Intent intent = new Intent(this, RegisterActivity.class);
            startActivityForResult(intent, REGISTER_REQUEST);
        }
        else if (v == loginButton) {
            String username = usernameInput.getText().toString();
            String password = passwordInput.getText().toString();

            viewModel.login(username, password);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REGISTER_REQUEST && resultCode == RESULT_OK && data != null) {
            String username = data.getStringExtra("username");
            String password = data.getStringExtra("password");
            usernameInput.setText(username);
            passwordInput.setText(password);

            viewModel.login(username, password);
        }
    }
}

File: app/src/main/java/com/tavro/parslie/outstanding/ui/view/LoginActivity.kt
package com.tavro.parslie.outstanding.ui.view

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.google.android.material.snackbar.Snackbar
import com.tavro.parslie.outstanding.BuildConfig
import com.tavro.parslie.outstanding.R
import com.tavro.parslie.outstanding.data.repository.PreferenceRepository
import com.tavro.parslie.outstanding.databinding.ActivityLoginBinding
import com.tavro.parslie.outstanding.ui.viewmodel.ContextualViewModelFactory
import com.tavro.parslie.outstanding.ui.viewmodel.LoginViewModel
import com.tavro.parslie.outstanding.util.Status

class LoginActivity : AppCompatActivity() {
    private lateinit var binding: ActivityLoginBinding
    private lateinit var viewModel: LoginViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_login)
        binding.loginVersion.text = getString(R.string.app_version, BuildConfig.VERSION_NAME)

        initViewModel()
        initListeners()
    }

    private fun initViewModel() {
        val viewModelFactory = ContextualViewModelFactory(this)
        viewModel = ViewModelProvider(this, viewModelFactory)[LoginViewModel::class.java]

        viewModel.getLoginResponse().observe(this, {
            when(it.status) {
                Status.LOADING -> {
                    binding.loginProgressBar.visibility = View.VISIBLE
                    binding.loginLoginBtn.isEnabled = false
                }
                Status.SUCCESS -> {
                    val prefs = PreferenceRepository(applicationContext)
                    prefs.authToken = it.data!!.token
                    prefs.authID = it.data.user.id

                    val intent = Intent(this, MainActivity::class.java)
                    startActivity(intent)
                    finish()
                }
                Status.ERROR -> {
                    binding.loginProgressBar.visibility = View.GONE
                    binding.loginLoginBtn.isEnabled = true

                    // TODO: implement error model client-side and server-side (display like "404 not found")
                    Snackbar.make(this, binding.root, "There was an error...", Snackbar.LENGTH_LONG).show()
                }
            }
        })
    }

    private fun validateEmail(email: String): Boolean {
        return when (email) {
            "" -> {
                binding.loginEmail.error = resources.getString(R.string.no_email_input)
                false
            }
            else -> true
        }
    }

    private fun validatePassword(password: String): Boolean {
        return when (password) {
            "" -> {
                binding.loginPassword.error = resources.getString(R.string.no_password_input)
                false
            }
            else -> true
        }
    }

    private fun initListeners() {
        binding.loginLoginBtn.setOnClickListener {
            val email = binding.loginEmail.text.toString()
            val password = binding.loginPassword.text.toString()

            val validEmail = validateEmail(email)
            val validPassword = validatePassword(password)

            if (validEmail && validPassword)
                viewModel.login(email, password)
        }

        binding.loginRegisterBtn.setOnClickListener {
            val intent = Intent(this, RegisterActivity::class.java)
            registerResults.launch(intent)
        }
    }

    private val registerResults = registerForActivityResult(StartActivityForResult()) {
        when(it.resultCode) {
            Activity.RESULT_OK -> {
                val email = it.data!!.getStringExtra("email")!!
                val password = it.data!!.getStringExtra("password")!!
                viewModel.login(email, password)
            }
        }
    }
}